File: app/routers/tasks.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from app.database import get_db
from app.models import Task, Project
from app.schemas import TaskCreate, Task as TaskSchema

router = APIRouter(prefix="/tasks", tags=["Tasks"])


@router.post("/", response_model=TaskSchema, status_code=status.HTTP_201_CREATED)
def create_task(
    task: TaskCreate,
    db: Session = Depends(get_db)
):
    project = db.query(Project).filter(Project.id == task.project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    db_task = Task(**task.model_dump())
    db.add(db_task)
    db.commit()
    db.refresh(db_task)
    return db_task


@router.get("/", response_model=List[TaskSchema])
def list_tasks(
    project_id: int = None,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    query = db.query(Task)
    if project_id:
        query = query.filter(Task.project_id == project_id)
    tasks = query.offset(skip).limit(limit).all()
    return tasks


@router.get("/{task_id}", response_model=TaskSchema)
def get_task(
    task_id: int,
    db: Session = Depends(get_db)
):
    task = db.query(Task).filter(Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.put("/{task_id}", response_model=TaskSchema)
def update_task(
    task_id: int,
    task_update: TaskCreate,
    db: Session = Depends(get_db)
):
    db_task = db.query(Task).filter(Task.id == task_id).first()
    if not db_task:
        raise HTTPException(status_code=404, detail="Task not found")
    
    for key, value in task_update.model_dump().items():
        setattr(db_task, key, value)
    
    db.commit()
    db.refresh(db_task)
    return db_task


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_task(
    task_id: int,
    db: Session = Depends(get_db)
):
    db_task = db.query(Task).filter(Task.id == task_id).first()
    if not db_task:
        raise HTTPException(status_code=404, detail="Task not found")
    
    db.delete(db_task)
    db.commit()
    return None

File: frontend/src/api/tasks.ts
import client from './client'
import { Task } from './types'

export const tasksAPI = {
  list: async (projectId?: number, skip: number = 0, limit: number = 100) => {
    const response = await client.get<Task[]>('/tasks/', {
      params: { project_id: projectId, skip, limit }
    })
    return response.data
  },

  get: async (taskId: number) => {
    const response = await client.get<Task>(`/tasks/${taskId}`)
    return response.data
  },

  create: async (data: Omit<Task, 'id' | 'created_at' | 'updated_at'>) => {
    const response = await client.post<Task>('/tasks/', data)
    return response.data
  },

  update: async (taskId: number, data: Omit<Task, 'id' | 'created_at' | 'updated_at'>) => {
    const response = await client.put<Task>(`/tasks/${taskId}`, data)
    return response.data
  },

  delete: async (taskId: number) => {
    await client.delete(`/tasks/${taskId}`)
  }
}
